compiler/rustc_middle/src/dep_graph/dep_node.rs

```diff
@@ -58,18 +58,128 @@ use crate::traits::query::{
     CanonicalTypeOpAscribeUserTypeGoal, CanonicalTypeOpEqGoal, CanonicalTypeOpNormalizeGoal,
     CanonicalTypeOpProvePredicateGoal, CanonicalTypeOpSubtypeGoal,
 };
+use crate::ty::query::{on_disk_cache, Query, QueryStats};
 use crate::ty::subst::{GenericArg, SubstsRef};
 use crate::ty::{self, ParamEnvAnd, Ty, TyCtxt};
 
 use rustc_data_structures::fingerprint::Fingerprint;
 use rustc_hir::def_id::{CrateNum, DefId, LocalDefId, CRATE_DEF_INDEX};
 use rustc_hir::definitions::DefPathHash;
 use rustc_hir::HirId;
+use rustc_serialize::{Decodable, Decoder, Encodable, Encoder};
 use rustc_span::symbol::Symbol;
+use rustc_span::DUMMY_SP;
 use std::hash::Hash;
 
 pub use rustc_query_system::dep_graph::{DepContext, DepNodeParams};
 
+pub type DepKind = &'static DepKindStruct;
+pub struct DepKindStruct {
+    pub(super) index: DepKindIndex,
```

Comment:
Could the fields be public so you can define your own DepKind outside rustc?

Reply:
This whole file should be moved outside `rustc_middle` eventually. However, there should be an exact correspondence between `DepKind`s, `DepKindIndex` and `DEP_KINDS` for encoding / decoding.